graph/bolt2/indexing.go

```diff
@@ -0,0 +1,519 @@
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	for i := 0; i < 256; i++ {
+		err := qs.db.Update(func(tx *bolt.Tx) error {
+			var err error
+			for j := 0; j < 256; j++ {
+				_, err = tx.CreateBucket(bucketFor(byte(i), byte(j)))
+				if err != nil {
+					return fmt.Errorf("could not create subbucket %d %d : %s", i, j, err)
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func bucketFor(i, j byte) []byte {
+	return []byte{'v', i, j}
+}
+
+func (qs *QuadStore) writeHorizonAndSize(tx *bolt.Tx) error {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.LittleEndian, qs.size)
+	if err != nil {
+		clog.Errorf("Couldn't convert size!")
+		return err
+	}
+	b := tx.Bucket(metaBucket)
+	werr := b.Put([]byte("size"), buf.Bytes())
+	if werr != nil {
+		clog.Errorf("Couldn't write size!")
+		return werr
+	}
+	buf.Reset()
+	err = binary.Write(buf, binary.LittleEndian, qs.horizon)
+
+	if err != nil {
+		clog.Errorf("Couldn't convert horizon!")
+	}
+
+	werr = b.Put([]byte("horizon"), buf.Bytes())
+
+	if werr != nil {
+		clog.Errorf("Couldn't write horizon!")
+		return werr
+	}
+	return err
+}
+
+func (qs *QuadStore) ApplyDeltas(deltas []graph.Delta, ignoreOpts graph.IgnoreOpts) error {
+	tx, err := qs.db.Begin(true)
+	tx.Bucket(logIndex).FillPercent = 0.9
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+nextDelta:
+	for _, d := range deltas {
+		link := proto.Primitive{}
+		mustBeNew := false
+		for _, dir := range quad.Directions {
+			val := d.Quad.Get(dir)
+			if val == nil {
+				continue
+			}
+			v := qs.resolveQuadValue(tx, val)
+			if v == 0 {
+				// Not found
+				if d.Action == graph.Delete {
+					if ignoreOpts.IgnoreMissing {
+						continue nextDelta
+					}
+					return fmt.Errorf("Deleting unknown quad: %s", d.Quad)
+				}
+				node, err := qs.createNodePrimitive(val)
+				if err != nil {
+					return err
+				}
+				qs.horizon++
+				node.ID = uint64(qs.horizon)
+				err = qs.index(tx, node, val)
+				mustBeNew = true
+				if err != nil {
+					return err
+				}
+				v = node.ID
+			}
+			link.SetDirection(dir, v)
+		}
+		if d.Action == graph.Delete {
+			id, err := qs.hasPrimitive(tx, &link)
+			if err != nil {
+				return err
+			}
+			err = qs.markAsDead(tx, id)
+			if err != nil {
+				return err
+			}
+			writePrimToBuf(&link)
+			qs.exists.TestAndRemove(bloomBuf)
+			qs.size--
+			continue
+		}
+
+		// Check if it already exists.
+		if !mustBeNew {
+			id, err := qs.hasPrimitive(tx, &link)
+			if err != nil {
+				return err
+			}
+			if id != 0 {
+				if ignoreOpts.IgnoreDup {
+					continue
+				}
+				return fmt.Errorf("adding duplicate link %v", d)
+			}
+		}
+		qs.horizon++
+		link.ID = uint64(qs.horizon)
+		link.Timestamp = time.Now().UnixNano()
+		err = qs.index(tx, &link, nil)
+		writePrimToBuf(&link)
+		qs.exists.Add(bloomBuf)
+		if err != nil {
+			return err
+		}
+		qs.size++
+	}
+	err = qs.flushMapBucket(tx)
```

Comment:
This map has the same scope as ApplyDelta, if I'm not mistaken. Make it a local variable here?